include/range/v3/utility/addressof.hpp

```diff
@@ -0,0 +1,40 @@
+// Range v3 library
+//
+//  Copyright Eric Niebler 2013-present
+//
+//  Use, modification and distribution is subject to the
+//  Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at
+//  http://www.boost.org/LICENSE_1_0.txt)
+//
+// Project home: https://github.com/ericniebler/range-v3
+//
+#ifndef RANGES_V3_UTLITY_ADDRESSOF_HPP
+#define RANGES_V3_UTLITY_ADDRESSOF_HPP
+
+#ifdef __cpp_lib_addressof_constexpr
+#include <memory>
+#endif
+
+namespace ranges
+{
+    /// \cond
+    namespace detail
+    {
+        template <typename T>
+        constexpr T* addressof(T& arg) noexcept
```

Comment:
It should be possible to detect at compile-time whether `T` has an overloaded `operator&` and fall back to a non-`constexpr` implementation that does the `reinterpret_cast` dance; I would prefer that than using `&arg` and silently getting some weird result.

Reply:
I can't seem to come up with a way that works on gcc C++11. I can't just `auto` deduce the return type of `my::test::addressof`, and if I add either a trailing or leading `decltype` expression that does `&t` or whatever, it doesn't find the `ignore operator&(ignore)` -- it just unconditionally gives me `T*`.

It looks like any unevaluated `&x` just short-circuits to doing addressof.

I can get this something like this working on gcc c++14, or clang.

Reply:
Never mind, I think I got something working.

Reply:
... and lol, range-v3 doesn't support C+11 anymore! 🎉 Well, I feel like I accomplished something anyway.